bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/Desugar.scala

```diff
@@ -187,7 +187,29 @@ object Desugar extends Transformer with TreeDSL {
       treeCopy.MatchStatement(expr, value,
           clauses map(c => matchExpressionClauseToBindStatement(result, c)),
           exprToBindStatement(result, elseExpression))
-    
+
+    case Record(label, fields) =>
+      var tailExpression: Option[(Variable, Expression)] = None
+      def defer(expr: Expression) = {
+        val newVar = Variable.newSynthetic()
+        tailExpression = Some((newVar, expr))
+        newVar
+      }
+
+      val newFields = fields.reverse.map { field =>
+        if (!tailExpression.isDefined && hasSideEffects(field))
```

Comment:
I think this would read better as `tailExpression.isEmpty && hasSideEffect(field)`